Source/Diagnostics/BoostedFrameDiagnostic.H

```diff
@@ -13,116 +13,231 @@
-    amrex::Real dt_snapshots_lab_;
-    amrex::Real dt_boost_;
-    int N_snapshots_;
-    int boost_direction_;
+    amrex::Real dx_;
+    amrex::Real dy_;
 
-    // For back-transformed diagnostics of grid fields, data_buffer_[i]
+    int ncomp_to_dump_;
+    std::vector<std::string> mesh_field_names_;
+
+    int file_num;
+    int initial_i;
+
+    // For back-transformed diagnostics of grid fields, data_buffer_
     // stores a buffer of the fields in the lab frame (in a MultiFab, i.e.
     // with all box data etc.). When the buffer if full, dump to file.
-    amrex::Vector<std::unique_ptr<amrex::MultiFab> > data_buffer_;
+    std::unique_ptr<amrex::MultiFab> data_buffer_;
     // particles_buffer_ is currently blind to refinement level.
-    // particles_buffer_[i][j] is a WarpXParticleContainer::DiagnosticParticleData where
-    // - i is the back-transformed snapshot number
-    // - j is the species number
-    amrex::Vector<amrex::Vector<WarpXParticleContainer::DiagnosticParticleData> > particles_buffer_;
+    // particles_buffer_[j] is a WarpXParticleContainer::DiagnosticParticleData
+    //  where - j is the species number for the current diag
+    amrex::Vector<WarpXParticleContainer::DiagnosticParticleData> particles_buffer_;
+    // buff_counter_ is the number of z slices in data_buffer_
+    int buff_counter_;
     int num_buffer_ = 256;
-    int max_box_size_ = 256;
-    // buff_counter_[i] is the number of z slices in data_buffer_[i]
-    // for snapshot number i.
-    amrex::Vector<int> buff_counter_;
+    int max_box_size = 256;
+    void updateCurrentZPositions(amrex::Real t_boost, amrex::Real inv_gamma,
+                                 amrex::Real inv_beta);
 
-    amrex::Vector<LabSnapShot> snapshots_;
+    void createLabFrameDirectories();
 
-    void writeParticleData(const WarpXParticleContainer::DiagnosticParticleData& pdata,
-                           const std::string& name, const int i_lab);
+    void writeLabFrameHeader();
+
+    /// Back-transformed lab-frame field data is copied from
+    /// tmp_slice to data_buffer where it is stored.
+    /// For the full diagnostic, all the data in the MultiFab is copied.
+    /// For the reduced diagnostic, the data is selectively copied if the
+    /// extent of the z_lab multifab intersects with the user-defined sub-domain
+    /// for the reduced diagnostic (i.e., a 1D, 2D, or 3D region of the domain).
+    virtual void AddDataToBuffer(amrex::MultiFab& tmp_slice_ptr, int i_lab,
+                 amrex::Gpu::ManagedDeviceVector<int> map_actual_fields_to_dump){};
+
+    /// Back-transformed lab-frame particles is copied from
+    /// tmp_particle_buffer to particles_buffer.
+    /// For the full diagnostic, all the particles are copied,
+    /// while for the reduced diagnostic, particles are selectively
+    /// copied if their trajectory crosses the user-defined
+    /// sub-domain for the reduced slice diagnostic.
+    virtual void AddPartDataToParticleBuffer(
```

Comment:
Is the description above up-to-date? To my understanding, particles are copied if they were in the use-defined sub-domain. Is that correct? If so, would it be more accurate to say
```
    /// For the full diagnostic, all the particles are copied,
    /// while for the reduced diagnostic, particles are selectively
    /// copied if their are inside the user-defined
    /// sub-domain for the reduced slice diagnostic.
```
?

Reply:
Ah yes. This is correct. I have now changed it as follows :
    /// For the full diagnostic, all the particles are copied,
    /// while for the reduced diagnostic, particles are selectively
    /// copied if their position in within the user-defined
    /// sub-domain +/- 1 cell size width for the reduced slice diagnostic.